src/partitioning.c

```diff
@@ -99,20 +179,90 @@ PGDLLEXPORT Datum get_partition_for_key(PG_FUNCTION_ARGS);
 
 PG_FUNCTION_INFO_V1(get_partition_for_key);
 
+/*
+ * Deprecated function to calculate partition hash values.
+ *
+ * This function assumes text input only.
+ */
 Datum
 get_partition_for_key(PG_FUNCTION_ARGS)
 {
-	struct varlena *data;
+	struct varlena *data = PG_GETARG_VARLENA_PP(0);
 	uint32		hash_u;
 	int32		res;
 
-	data = PG_GETARG_VARLENA_PP(0);
-
 	hash_u = DatumGetUInt32(hash_any((unsigned char *) VARDATA_ANY(data),
-						VARSIZE_ANY_EXHDR(data)));
+									 VARSIZE_ANY_EXHDR(data)));
 
 	res = (int32) (hash_u & 0x7fffffff);		/* Only positive numbers */
 
 	PG_FREE_IF_COPY(data, 0);
 	PG_RETURN_INT32(res);
 }
+
+PGDLLEXPORT Datum get_partition_hash(PG_FUNCTION_ARGS);
+
+PG_FUNCTION_INFO_V1(get_partition_hash);
+
+/*
+ * Compute a partition hash value for any input type.
+ *
+ * get_partition_hash() takes a single argument of anyelement type. We compute
+ * the hash based on the argument type information that we expect to find in the
+ * function expression in the function call context. If no such expression
+ * exists, or the type cannot be resolved from the expression, the function
+ * throws an error.
+ */
+Datum
+get_partition_hash(PG_FUNCTION_ARGS)
+{
+	Datum		arg = PG_GETARG_DATUM(0);
+	TypeCacheEntry *tce = fcinfo->flinfo->fn_extra;
+	Datum		hash;
+	FuncExpr   *fe;
+	Node	   *node;
+	Oid			argtype;
+	int32		res;
+
+	if (PG_NARGS() != 1)
+		elog(ERROR, "Unexpected number of arguments to partitioning function");
+
+	/* Get the function expression from the call info */
+	fe = (FuncExpr *) fcinfo->flinfo->fn_expr;
+
+	if (NULL == fe || !IsA(fe, FuncExpr))
+		elog(ERROR, "No function expression set when invoking partitioning function");
+
+	if (list_length(fe->args) != 1)
+		elog(ERROR, "Unexpected number of arguments in function expression");
+
+	node = linitial(fe->args);
+
+	switch (nodeTag(node))
+	{
+		case T_Var:
+			argtype = ((Var *) node)->vartype;
+			break;
+		case T_Const:
+			argtype = ((Const *) node)->consttype;
+			break;
+		case T_CoerceViaIO:
+			argtype = ((CoerceViaIO *) node)->resulttype;
+			break;
+		default:
+			elog(ERROR, "Unsupported expression argument node type %u", nodeTag(node));
+	}
+
+	if (tce == NULL)
+		tce = lookup_type_cache(argtype, TYPECACHE_HASH_FLAGS);
+
+	if (tce->hash_proc == InvalidOid)
```

Comment:
Why are we not falling back to hash_any on the text representation here as before?

Reply:
We use PostgreSQL's lookup function that supports virtually any Postgres type, as shown in the regression test for hashing in this PR. I think going beyond those supported in PostgreSQL has undefined behavior and I'd rather fail this than silently doing something weird that we are later stuck with. If users are complaining that they are trying to partition on a type that we do not support, we can add explicit support for it later. For custom types, you can either use your own partitioning function or create a hash function for your type that would resolve via `lookup_type_cache()`.

Reply:
Ok, I guess that makes more sense.